### PR TITLE
Rework ln invoice tags

### PR DESCRIPTION
**Pull request by Christewart**
This PR aims to rework the tag portion of the LN invoices. This is mostly restructing at this point, and moving serialization logic out of the abstract class `LnInvoiceTags` into the actual individual tags themselves.

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/util/Bech32.scala [hunk 1]
```diff
@@ -110,27 +113,60 @@ sealed abstract class Bech32 {
...
+
+  /** Takes a bech32 (base 5) byte array and encodes it to a string */
+  def encodeBase5ToString(b: Vector[UInt8]): String = {
+    require(!b.exists(_ >= u8ThirtyTwo), s"Number out of bech32 range in vector $b")
     b.map(b => charset(b.toInt)).mkString
   }
 
   /** Converts a byte vector from base 8 to base 5 */
-  def fromBase8ToBase5(bytes: ByteVector): Try[Vector[UInt8]] = {
+  def fromBase8ToBase5(bytes: ByteVector): Vector[UInt8] = {
```
Comment (Christewart) [inline core/src/main/scala/org/bitcoins/core/util/Bech32.scala]: note all of these names are misleding, it isn't really base8. What it really should be named to is something like 

`byteTo5bit`

**Comment by Christewart** [comment]
NOTE: Unit tests do not pass and the serialization scheme is still wrong on this PR.

File: core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala [hunk 2]
```diff
@@ -1,98 +1,129 @@
...
 
-  def fromStringToBech32(prefix: LnTagPrefix, tag: String): String = {
-    if (!tag.isEmpty) {
-      val uInt8Array = tag.map(a => a.toByte).map(i => UInt8(toUnsignedByte(i).toShort))
-      val base32Array = Bech32Address.encode(uInt8Array)
-      val bech32String = Bech32Address.encodeToString(base32Array.get)
-      val bech32DataLength = bech32EncodeDataLength(bech32String)
-      prefix.value + bech32DataLength + bech32String
-    } else { "" }
+  case class ScriptPubKeyTag(spk: ScriptPubKey) extends LnInvoiceTag {
```
Comment (TannerR1776) [inline core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala]: This is meant to be the `n` tag which denotes the "33-byte public key of the payee node." I think you are getting it confused with the `f` tag which denotes the "Fallback on-chain address: for bitcoin, this starts with a 5-bit version and contains a witness program or P2PKH or P2SH address." ?

File: core/src/main/scala/org/bitcoins/core/protocol/ln/LnTagPrefix.scala [hunk 3]
```diff
@@ -19,7 +19,7 @@ object LnTagPrefix {
     override def value: String = "d"
   }
 
-  case object SignaturePubKey extends LnTagPrefix {
+  case object ScriptPubKey extends LnTagPrefix {
```
Comment (TannerR1776) [inline core/src/main/scala/org/bitcoins/core/protocol/ln/LnTagPrefix.scala]: Again, this is for tag `n` which denotes the "33-byte public key of the payee node". The name SignaturePubKey probably comes from this line in BOLT-11: "A writer MAY include one n field, which MUST be set to the public key used to create the signature." We could probably rename to something more intuitive?

File: core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala [hunk 4]
```diff
@@ -1,98 +1,129 @@
...
+  def prefix: LnTagPrefix
 
-  def spk: Option[ScriptPubKey]
+  /** Converts the data portion of the LnInvoiceTag to a bech32 string */
+  def dataToBech32: String = {
+    Bech32.encode8bitToString(bytes)
+  }
 
-  def descriptionHash: Option[Sha256Digest]
+  override def toString: String = {
```
Comment (TannerR1776) [inline core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala]: I like splitting out the Tags into their own objects, and giving each their own serialization / `toString` capabilities but, as far as I can tell, none of the encoding used in the `LnInvoiceTag` objects is correct. Unfortunately, it isn't as simple as just converting the data to bytes and Bech32 encoding it. We will need to go back and revise.

**Review by TannerR1776 — COMMENTED** [review]
I have reviewed the code. Overall, most changes are good and tests are passing. I left just a few questions for more detail, but I am fine with merging as is. I did note, we will need to go back and re-add serialization for `LnInvoice`'s

File: core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala [hunk 5]
```diff
@@ -105,23 +174,23 @@ sealed abstract class LnInvoiceTags {
...
-    b.append(fallbackAddress.getOrElse(empty))
+    b.append(description.map(_.toString).getOrElse(empty))
+    b.append(nodeId.map(_.toString).getOrElse(empty))
+    b.append(descriptionHash.map(_.toString).getOrElse(empty))
+    b.append(expiryTime.map(_.toString).getOrElse(empty))
+    b.append(cltvExpiry.map(_.toString).getOrElse(empty))
+    b.append(fallbackAddress.map(_.toString).getOrElse(empty))
 
-    //TODO: add routing info
+    val routes = routingInfo.map(_.toString).getOrElse(empty)
```
Comment (TannerR1776) [inline core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala]: Is there a reason these two lines aren't just combined into `b.append(routingInfo.map(_.toString).getOrElse(empty))`?

File: core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala [hunk 6]
```diff
@@ -138,7 +207,7 @@ sealed abstract class LnInvoiceTags {
     val quotient = Bech32EncodedString.length / 32
     val remainder = Bech32EncodedString.length % 32
     val v = Vector(UInt8(quotient.toShort), UInt8(remainder.toShort))
-    Bech32.encode8bitToString(v)
+    Bech32.encode5bitToString(v)
```
Comment (TannerR1776) [inline core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala]: This function is orphaned, and should be removed. Functionality has been moved to `dataLength` and `encodeNumber`.

File: core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala [hunk 7]
```diff
@@ -66,23 +96,62 @@ object LnInvoiceTag {
...
+      Bech32.encode5bitToString(bytes)
+    }
   }
 
   case class CltvExpiryTag(u64: UInt64) extends LnInvoiceTag {
     override val prefix: LnTagPrefix = LnTagPrefix.CltvExpiry
 
-    override val bytes: ByteVector = u64.bytes
+    override val bytes: ByteVector = {
+      encodeNumber(u64.toInt)
```
Comment (TannerR1776) [inline core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala]: This isnt safe unless we add some `require` statements to the the` CtlvExpiryTag` and `ExpiryTimeTag`.

**Review by TannerR1776 — COMMENTED** [review]
A lot of changes to `LnInvoiceTags`, but the tests are passing and the serialization looks good. I left a few comments.

File: core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala (hunk 5 above)
Comment (Christewart) [reply core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala]: It gets a little long. I like using intermediate `vals` if the thing we are passing to a method is not trivial.

**Comment by TannerR1776** [comment]
Awesome stuff!

**Pull request merged**